GoTest doesn't show quickfix window on failed Example tests

### What did you do? (required: The issue will be **closed** when not provided)
Given the following source code:
```
package factorial

// Factorial calculates the factorial of a given number
func Factorial(n int) int {
	if n <= 1 {
		return 1
	}

	return n * Factorial(n-1)
}
```

and the following test code:
```
package factorial_test

import (
	"fmt"

	"github.com/preslavmihaylov/go-playground/factorial"
)

func ExampleFactorial() {
	fmt.Println(factorial.Factorial(5))
	// Output:
	// 121
}
```

I ran `:GoTest`

### What did you expect to happen?
The test fails (this happened)
and the quickfix window opens with the failing function (didn't happen)
### What happened instead?
Instead, I just get a vague error message:
`vim-go: FAIL        github.com/preslavmihaylov/go-playground/factorial^I0.001s`

If I run `go test` from the terminal manually, this is what I get:
```
--- FAIL: ExampleFactorial (0.00s)
got:
120
want:
121
FAIL
exit status 1
FAIL    github.com/preslavmihaylov/go-playground/factorial      0.001s
```

Also, if there are normal tests in addition to this example test case, the quickfix window doesn't open at all even for the normal tests.

### Configuration (**MUST** fill this out):

#### vim-go version:
https://github.com/fatih/vim-go/tree/3e9a29279627d1ac50a855376b5da4dc0f809356

####  `vimrc` you used to reproduce (use a *minimal* vimrc with other plugins disabled; do not link to a 2,000 line vimrc):
<details><summary>vimrc</summary><br><pre>
""" -------------------- vim-go -------------------------
" all errors are shown in the quickfix window
let g:go_list_type = "quickfix"

" automatically add import paths when saving file
"let g:go_fmt_command = "goimports"

" enable syntax highlighting
let g:go_highlight_types = 1
let g:go_highlight_fields = 1
let g:go_highlight_functions = 1
let g:go_highlight_function_calls = 1
let g:go_highlight_operators = 1
let g:go_highlight_extra_types = 1
let g:go_highlight_build_constraints = 1
let g:go_highlight_generate_tags = 1

" enable go metalinter
let g:go_metalinter_enabled = ['vet', 'golint', 'errcheck']
let g:go_metalinter_autosave = 1

" automatically show GoInfo output
" let g:go_auto_type_info = 1

" automatically highlight variable your cursor is on
let g:go_auto_sameids = 1

" adjust quickfix window height
let g:go_list_height = 8

let g:go_term_mode = "split above"
</pre></details>

[link to full vimrc](https://github.com/preslavmihaylov/dotfiles/blob/master/vimrc)

#### Vim version (first three lines from `:version`):
NVIM v0.3.4

####  Go version (`go version`):
1.11.4 linux/amd64

#### Go environment
<details><summary><code>go env</code> Output:</summary><br><pre>
GOARCH="amd64"
GOBIN=""
GOCACHE="/home/pmihaylov/.cache/go-build"
GOEXE=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOOS="linux"
GOPATH="/home/pmihaylov/programming/go"
GOPROXY=""
GORACE=""
GOROOT="/usr/local/go"
GOTMPDIR=""
GOTOOLDIR="/usr/local/go/pkg/tool/linux_amd64"
GCCGO="gccgo"
CC="gcc"
CXX="g++"
CGO_ENABLED="1"
GOMOD=""
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build501012645=/tmp/go-build -gno-record-gcc-switches"
</pre></details>

--- a/autoload/go/test-fixtures/test/src/example/example_test.go
+++ b/autoload/go/test-fixtures/test/src/example/example_test.go
@@ -1,0 +1,10 @@
+package main
+
+import (
+	"fmt"
+)
+
+func ExampleHelloWorld() {
+	fmt.Println("Hello, World")
+	// Output: What's shakin
+}

--- a/autoload/go/test.vim
+++ b/autoload/go/test.vim
@@ -166,9 +166,17 @@ function! s:errorformat() abort
   let format .= ",%-G" . indent . "%#--- PASS: %.%#"
 
   " Match failure lines.
-  "
+
+  " Example failures start with '--- FAIL: ', followed by the example name
+  " followed by a space , followed by the duration of the example in
+  " parantheses. They aren't nested, though, so don't check for indentation.
+  " The errors from them also aren't indented and don't report file location
+  " or line numbers, so those won't show up. This will at least let the user
+  " know which example failed, though.
+  let format .= ',%G--- FAIL: %\\%(Example%\\)%\\@=%m (%.%#)'
+
   " Test failures start with '--- FAIL: ', followed by the test name followed
-  " by a space the duration of the test in parentheses
+  " by a space, followed by the duration of the test in parentheses.
   "
   " e.g.:
   "   '--- FAIL: TestSomething (0.00s)'

--- a/autoload/go/test_test.vim
+++ b/autoload/go/test_test.vim
@@ -66,9 +66,9 @@ endfunc
 func! Test_GoTestShowName() abort
   let expected = [
         \ {'lnum': 0, 'bufnr': 0, 'col': 0, 'valid': 1, 'vcol': 0, 'nr': -1, 'type': '', 'pattern': '', 'text': 'TestHelloWorld'},
-        \ {'lnum': 6, 'bufnr': 7, 'col': 0, 'valid': 1, 'vcol': 0, 'nr': -1, 'type': '', 'pattern': '', 'text': 'so long'},
+        \ {'lnum': 6, 'bufnr': 8, 'col': 0, 'valid': 1, 'vcol': 0, 'nr': -1, 'type': '', 'pattern': '', 'text': 'so long'},
         \ {'lnum': 0, 'bufnr': 0, 'col': 0, 'valid': 1, 'vcol': 0, 'nr': -1, 'type': '', 'pattern': '', 'text': 'TestHelloWorld/sub'},
-        \ {'lnum': 9, 'bufnr': 7, 'col': 0, 'valid': 1, 'vcol': 0, 'nr': -1, 'type': '', 'pattern': '', 'text': 'thanks for all the fish'},
+        \ {'lnum': 9, 'bufnr': 8, 'col': 0, 'valid': 1, 'vcol': 0, 'nr': -1, 'type': '', 'pattern': '', 'text': 'thanks for all the fish'},
       \ ]
 
   let g:go_test_show_name=1
@@ -78,18 +78,25 @@ endfunc
 
 func! Test_GoTestVet() abort
   let expected = [
-        \ {'lnum': 6, 'bufnr': 10, 'col': 2, 'valid': 1, 'vcol': 0, 'nr': -1, 'type': '', 'pattern': '', 'text': 'Errorf format %v reads arg #1, but call has 0 args'},
+        \ {'lnum': 6, 'bufnr': 11, 'col': 2, 'valid': 1, 'vcol': 0, 'nr': -1, 'type': '', 'pattern': '', 'text': 'Errorf format %v reads arg #1, but call has 0 args'},
       \ ]
   call s:test('veterror/veterror.go', expected)
 endfunc
 
 func! Test_GoTestTestCompilerError() abort
   let expected = [
-        \ {'lnum': 10, 'bufnr': 8, 'col': 16, 'valid': 1, 'vcol': 0, 'nr': -1, 'type': '', 'pattern': '', 'text': 'cannot use r (type struct {}) as type io.Reader in argument to ioutil.ReadAll:'},
+        \ {'lnum': 10, 'bufnr': 9, 'col': 16, 'valid': 1, 'vcol': 0, 'nr': -1, 'type': '', 'pattern': '', 'text': 'cannot use r (type struct {}) as type io.Reader in argument to ioutil.ReadAll:'},
         \ {'lnum': 0, 'bufnr': 0, 'col': 0, 'valid': 1, 'vcol': 0, 'nr': -1, 'type': '', 'pattern': '', 'text': 'struct {} does not implement io.Reader (missing Read method)'}
       \ ]
 
   call s:test('testcompilerror/testcompilerror_test.go', expected)
+endfunc
+
+func! Test_GoTestExample() abort
+  let expected = [
+        \ {'lnum': 0, 'bufnr': 0, 'col': 0, 'valid': 1, 'vcol': 0, 'nr': -1, 'type': '', 'pattern': '', 'text': 'ExampleHelloWorld'}
+      \ ]
+  call s:test('example/example_test.go', expected)
 endfunc
 
 func! s:test(file, expected, ...) abort

